### PR TITLE
chore: Saving dates overrides directly on change

### LINKED ISSUE #11191: [CAL-2410] Auto-save availability schedule right after a date override is added
**Problem**

*adding a date override in 'availbility' requires scrolling up and clicking 'save'. Our entire team have been caught out by this at one point or another!*

**Solution**

An ideal solution is that our entire app has auto-save on availability schedules, event types, settings etc. In the mean time, we need to at least save each availability schedule after you add a date override. So once the modal is closed, the schedule is saved. You do not have to also hit the save button.

We should throw a small toast notification saying "Date override added" once the modal is closed.

<sub>From [SyncLinear.com](https://synclinear.com) | [CAL-2410](https://linear.app/calcom/issue/CAL-2410/auot-save-availability-schedule-right-after-a-date-override-is-added)</sub>

### LINKED ISSUE #13140: [CAL-2980] Adding a date override requires two clicks
Surprised this is required since when you click on "Add override", the toast says "Override added", but then if you refresh the page it disappears. You need to click on the general "Save" button for it to work.

CleanShot 2024-01-09 at 19 19 45

<sub>From [SyncLinear.com](https://synclinear.com) | [CAL-2980](https://linear.app/calcom/issue/CAL-2980/adding-a-date-override-requires-two-clicks)</sub>

### PR DESCRIPTION
**Pull request by stylessh**
## What does this PR do?

This PR saves the dates overrides directly on change. Previously you have to click save button after saving the dates.

Fixes #13140, #11191  and related

https://github.com/calcom/cal.com/assets/45767683/be7575c4-5633-4c58-a211-b31bbd172f8c

## Requirement/Documentation

It's rather annoying to do an "extra' click to actually save the overrides.

Refer to #13140 to view more details about the issue.

## Type of change

- Bug fix (non-breaking change which fixes an issue)
- Chore (refactoring code, technical debt, workflow improvements)
- New feature (non-breaking change which adds functionality)

## How should this be tested?

- Go to availability page.
- Create a new date override.
- Try to save the override. 

## Mandatory Tasks

- [x] Make sure you have self-reviewed the code. A decent size PR without self-review might be rejected.

### CODE REVIEW

File: packages/features/schedules/components/invalidate.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import { type RouterContext } from "@calcom/trpc/react";
+
+export async function invalidateSchedules<T>(
```
Comment (stylessh) [inline packages/features/schedules/components/invalidate.ts:3]: Created a new util to invalidate and refresh the schedules data.

File: packages/trpc/react/trpc.ts [hunk 2]
```diff
@@ -132,3 +132,5 @@ export const transformer = superjson;
 
 export type RouterInputs = inferRouterInputs<AppRouter>;
 export type RouterOutputs = inferRouterOutputs<AppRouter>;
+
+export type RouterContext = Awaited<ReturnType<typeof trpc.useContext>>;
```
Comment (stylessh) [inline packages/trpc/react/trpc.ts:136]: new export to get tRPC context type.

File: packages/features/schedules/components/DateOverrideInputDialog.tsx [hunk 3]
```diff
@@ -132,8 +145,47 @@ const DateOverrideForm = ({
...
               })
             : datesInRanges
+        ).map((date) => ({
+          ranges: [date],
+        }));
+
+        if (scheduleId) {
+          const newSchedule = schedule as AvailabilityFormValues & { availability: Schedule };
+
+          await updateMutation.mutateAsync(
```
Comment (stylessh) [inline packages/features/schedules/components/DateOverrideInputDialog.tsx:155]: Updating on change async, this reflects the data and update directly on db.

**Comment by PeerRich** [comment]
btw. i wonder if we wanna keep the dialog open after adding too? 🤔 thoughts on that?

**Comment by stylessh** [comment]
It could be an option. Saving and cleaning the state but keeping the modal open in case the user wants to add more overrides (?)

**Review by Udit-takkar — COMMENTED** [review]
https://github.com/calcom/cal.com/assets/53316345/897ba369-9d92-4f59-8678-f903d82c792f

can you fix this Dropdown bug?

**Review by Udit-takkar — COMMENTED** [review]
Adding new overrides removed old overrides

https://github.com/calcom/cal.com/assets/53316345/526af28e-6af8-4489-bb7e-e37c4635eefa

**Comment by Udit-takkar** [comment]
Not able to edit saved override

https://github.com/calcom/cal.com/assets/53316345/fb7a6057-4562-42ea-bba6-f034e6ce4eb3

File: packages/features/schedules/components/DateOverrideInputDialog.tsx [hunk 4]
```diff
@@ -208,6 +257,7 @@ const DateOverrideInputDialog = ({
   Trigger: React.ReactNode;
   onChange: (newValue: TimeRange[]) => void;
   value?: TimeRange[];
+  schedule: AvailabilityFormValues;
```
Comment (Udit-takkar) [inline packages/features/schedules/components/DateOverrideInputDialog.tsx:260]: This is not used anywhere DateOverrideInputDialog component

Comment (Udit-takkar) [reply packages/features/schedules/components/DateOverrideInputDialog.tsx:260]: This is also causing type error

**Comment by stylessh** [comment]
We're mainly having some issues because of the availability API endpoint, it currently override all the information on update if we don't explicitly pass it.

**Comment by stylessh** [comment]
I'm noticing there's a strange behavior, when you select a date override and then select the day after (let's say 18th and 19th) each time range goes inside the first day (in this case 18th), leaving the 19th as blank. @PeerRich @Udit-takkar 

https://github.com/calcom/cal.com/assets/45767683/9e985457-7d83-4ed7-b553-530607c58422

**Comment by stylessh** [comment]
Should be fine now

https://github.com/calcom/cal.com/assets/45767683/08f42769-c050-4e98-8371-1d38d380d7d8

File: packages/trpc/server/routers/viewer/availability/schedule/get.handler.ts [hunk 5]
```diff
@@ -94,18 +94,27 @@ export const getHandler = async ({ ctx, input }: GetOptions) => {
...
-        (item) => yyyymmdd(item.ranges[0].start) === yyyymmdd(override.date!)
+
+      // Create a new date string for the current override
+      // early return prevents override.date from ever being empty.
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      const overrideDateString = yyyymmdd(override.date!);
+
+      // Check if there's a range with the day of the override date, to append a new range to it
+      const existingDayIndex = acc.findIndex(
+        (item) => yyyymmdd(dayjs(item.day)) === dayjs(override.date).format("YYYY-MM-DD")
```
Comment (stylessh) [inline packages/trpc/server/routers/viewer/availability/schedule/get.handler.ts:105]: This was causing a bug that would put the range time of a consecutive day in a previous one. For example, 18th 9:00AM - 5:00PM and 19th 9:00AM - 5:00PM would be after this 18th 9:00AM -5:00PM (x2), and 19th would have gone away.

**Comment by stylessh** [comment]
any updates?

**Comment by Udit-takkar** [comment]
> any updates?

I'll review the PR after v3.8 release as the milestone of this PR is 3.9.

**Comment by Udit-takkar** [comment]
@stylessh could you please resolve the conflicts?

**Comment by CarinaWolli** [comment]
Hey @stylessh, thank you for putting the work into that 🙏 But we decided that we need to fix this properly and this solution would result in a confusing UX. 

To explain in more detail: 
If the override will be saved after clicking 'add override' in the modal it would be very weird if updating a date override wouldn't also save because the UI for that is exactly the same. 
Sounds like 'let's just save after updating too', but then the delete button should automatically save the changes as well. We want to avoid this mix of some fields needing the 'save' button and some not.

**Pull request closed without merging**